### PR TITLE
br/pkg/storage: add oss support on s3 sdk

### LINKED ISSUE #34224: br/pkg/storage: s3 sdk support to access aliyun oss by ram type
## Feature Request

**Is your feature request related to a problem? Please describe:**

no
**Describe the feature you'd like:**

use s3 sdk, access aliyun oss by ram type: on ram ecs, can auto request the sts token ,then use sts token to access oss.
**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

the doc at aliyun , example for python : [基于临时凭证，使用Python SDK访问OSS](https://partners-intl.aliyun.com/help/zh/elastic-compute-service/latest/use-ram-roles-to-access-other-alibaba-cloud-services#section-fgc-wkp-1gq)

### PR DESCRIPTION
**Pull request by knull-cn**
### What problem does this PR solve?

Issue Number: close #34224

Problem Summary:

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

test case by sts token :
- on aliyun ecs 
  - has apply ram role ,and dumpling and lightning ok 
  - no ram role, dumpling failed
- on aws , apply iam role ,and dumpling and lightning ok

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: br/pkg/storage/s3.go [hunk 1]
```diff
@@ -241,7 +244,70 @@ func NewS3Storage( // revive:disable-line:flag-parameter
 	})
 }
 
-func newS3Storage(backend *backuppb.S3, opts *ExternalStorageOptions) (*S3Storage, error) {
+const (
+	ossRAMMetaUrl = "http://100.100.100.200/latest/meta-data/ram/security-credentials/"
+	ossRAMAK      = "AccessKeyId"
+	ossRAMSK      = "AccessKeySecret"
```
Comment (lichunzhu) [inline br/pkg/storage/s3.go:250]:
```suggestion
	ossRAMAccessKey      = "AccessKeyId"
	ossRAMSecretKey      = "AccessKeySecret"
```

File: br/pkg/storage/s3.go [hunk 2]
```diff
@@ -241,7 +244,70 @@ func NewS3Storage( // revive:disable-line:flag-parameter
...
+
+func autoNewCred(qs *backuppb.S3, opts *ExternalStorageOptions) (cred *credentials.Credentials, err error) {
+	if qs.AccessKey != "" && qs.SecretAccessKey != "" {
+		return credentials.NewStaticCredentials(qs.AccessKey, qs.SecretAccessKey, ""), nil
+	}
+	endpoint := qs.Endpoint
+	// if endpoint is empty,return no error and run default(aws) follow.
+	if endpoint == "" {
+		return nil, nil
+	}
```
Comment (lichunzhu) [inline br/pkg/storage/s3.go:266]: Seems that we can delete this part.

File: br/pkg/storage/s3.go [hunk 3]
```diff
@@ -241,7 +244,70 @@ func NewS3Storage( // revive:disable-line:flag-parameter
 	})
 }
 
-func newS3Storage(backend *backuppb.S3, opts *ExternalStorageOptions) (*S3Storage, error) {
+const (
+	ossRAMMetaUrl = "http://100.100.100.200/latest/meta-data/ram/security-credentials/"
```
Comment (lichunzhu) [inline br/pkg/storage/s3.go:248]: Are you sure about the URL `100.100.100.200`?

File: br/pkg/storage/s3.go [hunk 4]
```diff
@@ -241,7 +244,70 @@ func NewS3Storage( // revive:disable-line:flag-parameter
...
+	var mp map[string]string
+	err = json.Unmarshal(d, &mp)
+	if err != nil {
+		return nil, errors.Annotate(err, "createOssRamCred failed parse sts token")
+	}
+	ak := mp[ossRAMAK]
+	sk := mp[ossRAMSK]
+	token := mp[ossRAMToken]
+	return credentials.NewStaticCredentials(ak, sk, token), nil
+}
```
Comment (lichunzhu) [inline br/pkg/storage/s3.go:308]: Can this function be replaced by https://github.com/aliyun/alibaba-cloud-sdk-go/blob/eb79f24/sdk/auth/credentials/providers/instance_metadata.go#L16?

File: br/pkg/storage/s3.go (hunk 2 above)
Comment (knull-cn) [reply br/pkg/storage/s3.go:266]: delete this part is ok 
but ,maybe other cloud should parse more and more part code ,and it is wasted

File: br/pkg/storage/s3.go (hunk 4 above)
Comment (knull-cn) [reply br/pkg/storage/s3.go:308]: that is good idea 👍

**Review by lichunzhu — APPROVED** [review]
LGTM

**Comment by lichunzhu** [comment]
@3pointer PTAL

**Review by 3pointer — APPROVED** [review]
LGTM

**Pull request merged**